### ISSUE TITLE
Training a hierarchical multi-class classifier in stages using --initial_regressor

### ISSUE THREAD

**Issue by usptact** [JohnLangford/vowpal_wabbit#1391]
I have a fairly small dataset (~5K) of short text strings. The annotation has two levels of global annotation annotation: K classes in the top-level and K_i variable number of classes under each top-level class i.

The first idea is to simply build a flat multi-class classifier but the problem is the sparsity of the data. Some top-level and low-level class combinations have only a handful of examples. Such classifier will work badly.

Experiments have shown that training a multi-class classifier to discriminate the top-level classes works pretty good when evaluated on a held-out test set. No problems here. My question is how to train a classifier that also predicts the lower level labels as well.

Thinking in Bayesian terms, I would build a hierarchical partial pooling model. The weights for top-level classifier will be learned jointly with the weights of the lower level classifier. The respective lower level classifier weights are only influenced by the top-level classifier weights (used as a prior). The training data is used more efficiently.

In this context, I am thinking to do the following:
(1) Train the top-level classifier
(2) Train each lower level classifier separately BUT use the initial regressor option by providing the model learned in the step (1).

Is this going to work? Perhaps there are better alternatives?

UPDATE:
I noticed the option `--multilabel_oaa`. Will it work in my setting of relatively small dataset?

UPDATE2:
I gave it more thought and it seems that this might not be a good idea or simply won't work. In Bayesian settings updates goes both ways: from high level to low level, and vice versa. The upper level model is receiving updates from all lower level models; the lower level models are "shrinking" towards the upper level model depending of how much training data is available: the less data, the more the model is similar to the upper level. If more data is available at a particular lower level model, it has a chance to be quite different from the top level model.

**Comment by JohnLangford** [comment]
I don't believe step (2) will work effectively because initial regressor generally need to be of the type that you want.  Options:
(1) Create a custom reduction in VW to do this.   That might be my approach, but it involves C++ coding.  To go that way, take a look at multilabel_oaa.cc: https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/multilabel_oaa.cc
(2) There are several systems for doing logarithmic time prediction in VW (lomtree, recall tree), but they don't take into account a preexisting hierarchy. 
(3) There is some experimental code which we've found is particularly useful for many classes with few labels in @wensun's branch.  This is however quite experimental for now.
